mlir/include/mlir/IR/OpImplementation.h

```diff
@@ -1602,14 +1602,33 @@ class OpAsmParser : public AsmParser {
                   SmallVectorImpl<Value> &result) {
     size_t operandSize = llvm::range_size(operands);
     size_t typeSize = llvm::range_size(types);
-    if (operandSize != typeSize)
-      return emitError(loc)
-             << operandSize << " operands present, but expected " << typeSize;
+    if (operandSize == typeSize) {
+      for (auto [operand, type] : llvm::zip_equal(operands, types))
+        if (resolveOperand(operand, type, result))
+          return failure();
+      return success();
+    }
 
-    for (auto [operand, type] : llvm::zip_equal(operands, types))
-      if (resolveOperand(operand, type, result))
-        return failure();
-    return success();
+    InFlightDiagnostic diag = emitError(loc)
+                              << "number of operands and types do not match";
+    std::string lesserQuantityText = "operand";
+    if (operandSize != 1)
+      lesserQuantityText += "s";
+    std::string higherQuantityText = "type";
+    if (typeSize != 1)
+      higherQuantityText += "s";
+
+    size_t lesserQuantity = operandSize;
+    size_t higherQuantity = typeSize;
+    if (operandSize > typeSize) {
+      std::swap(lesserQuantity, higherQuantity);
+      std::swap(lesserQuantityText, higherQuantityText);
+    }
+
+    diag.attachNote() << "got " << higherQuantity << " " << higherQuantityText
+                      << " but only " << lesserQuantity << " "
+                      << lesserQuantityText;
```

Comment:
What do you think about emitting something simpler like `got N operands and K types`?

Reply:
I don't feel super strong about either options, so if you feel the current version is worse or not worth the extra implementation complexity I can gladly change it!

Reply:
Yeah I feel like the current version is a little bit overcomplicated for little benefit over what I suggested above

Reply:
I would also +1 keeping it simpler. I would also keep it as part of the main message and not a note.